lmdeploy/vl/model/phi3_vision.py

```diff
@@ -0,0 +1,206 @@
+            temp_sub_GN = self.sub_GN.repeat(1, h * 12, 1, 1)
+            sub_img = torch.cat([sub_img, temp_sub_GN],
+                                dim=2).reshape(1, -1, 4 * C)
+            # (1, num_img_tokens, 1024*4)
+
+            # glb + sub
+            if self.hd_transform_order == 'glb_sub':
+                output_imgs.append(
+                    torch.cat([glb_img, self.glb_GN, sub_img], dim=1))
+            elif self.hd_transform_order == 'sub_glb':
+                output_imgs.append(
+                    torch.cat([sub_img, self.glb_GN, glb_img], dim=1))
+            else:
+                raise NotImplementedError(
+                    f'hd_transform_order = {self.hd_transform_order}'
+                )  # noqa E501
+
+            temp_len = int((h * w + 1) * 144 + 1 + (h + 1) * 12)
+            assert temp_len == output_imgs[-1].shape[
+                1], f'temp_len: {temp_len}, output_imgs[-1].shape[1]: {output_imgs[-1].shape[1]}'  # noqa E501
+            output_len.append(temp_len)
+
+        img_set_tensor = []
+        for _output_img in output_imgs:
+            img_feature_proj = self.img_projection(
+                _output_img.to(target_device).to(target_dtype))
+            img_set_tensor.append(img_feature_proj)
+    elif img_embeds.ndim == 4:
+        tt = (self.get_img_features(img_embeds).to(target_device).to(
+            target_dtype).reshape(-1, self.image_dim_out))
+        img_set_tensor = self.img_projection(tt)  # adapted visual features.
+    elif img_embeds.ndim == 3:
+        tt = (img_embeds.to(target_device).to(target_dtype).view(
+            -1, self.image_dim_out))
+        img_set_tensor = self.img_projection(tt)  # adapted visual features.
+    else:
+        raise NotImplementedError
+    return img_set_tensor
+
+
+class Phi3VisionModel(VisonModel):
+    """Llava hf vision model."""
+
+    def __init__(self, model_path: str, with_llm: bool = False):
+        self.model_path = model_path
+        self.with_llm = with_llm
+        self.hf_config = AutoConfig.from_pretrained(model_path,
+                                                    trust_remote_code=True)
+        self.build_model()
+
+    def build_model(self):
+        from accelerate import init_empty_weights, load_checkpoint_and_dispatch
+        from accelerate.utils import get_balanced_memory, infer_auto_device_map
+
+        with init_empty_weights(), warnings.catch_warnings():
+            warnings.simplefilter('ignore')
+            from transformers import AutoModelForCausalLM
+            model = AutoModelForCausalLM.from_config(self.hf_config,
+                                                     trust_remote_code=True)
+            if not self.with_llm:
+                del model.lm_head
+                del model.model.layers
+                del model.model.norm
+                del model.model.embed_tokens
+                del model.model.vision_embed_tokens.wte
+                setattr(model, 'lm_head', None)
```

Comment:
Is it necessary after `del model.lm_head`?

Reply:
could be removed